### PR TITLE
MenuApi.cpp - Fixed invalidated iterator bug in MenuApi.cpp LuaContex…

### PR DESCRIPTION
**Pull request by CitizenKyle**
…t::update_menus() that Visual Studio caught with _ITERATOR_DEBUG_LEVEL = 2 (on by default in Debug mode)

  i.e., menus.erase(it--) when it == menus.begin()
  See http://stackoverflow.com/questions/16269696/erasing-while-iterating-an-stdlist.

Cleaning up warnings that get past /W4 in MSVC.

NonAnimatedRegions.cpp, Enemy.cpp, Sound.cpp, Video.cpp, PathFinding.cpp, GameApi.cpp, PathMovement.cpp, GameCommands.cpp, MapData.cpp, SPC_Filter
  Fixed warnings like this: warning C4456: declaration of 'foo' hides previous local declaration
                  and this: warning C4459: declaration of 'bar' hides global declaration
                  and this: warning C4457: declaration of 'baz' hides function parameter

### CODE REVIEW

File: src/third_party/snes_spc/SPC_Filter.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ void SPC_Filter::run( short* io, int count )
 			int delta = f - pp1;
 			pp1 = f;
 			int s = sum >> (gain_bits + 2);
-			sum += (delta * gain) - (sum >> bass);
+			sum += (delta * g) - (sum >> b);
 			
 			// Clamp to 16 bits
 			if ( (short) s != s )
```
Comment (christopho) [inline src/third_party/snes_spc/SPC_Filter.cpp]: SPC_Filter.cpp is third-party code, I prefer that we don't change it.

File: src/movements/PathMovement.cpp [hunk 2]
```diff
@@ -371,19 +371,19 @@ void PathMovement::snap() {
...
     // if we haven't started to move the entity towards an intersection of the grid, do it now
     set_snapping_trajectory(Point(x, y), Point(snapped_x, snapped_y));
     snapping = true;
-    stop_snapping_date = now + 500; // timeout
+    stop_snapping_date = now_1 + 500; // timeout
   }
   else {
     // the entity is currently trying to move towards the closest grid intersection
 
-    uint32_t now = System::now();
```
Comment (christopho) [inline src/movements/PathMovement.cpp]: Just remove this line, both "now" variables actually have the same value.

File: src/lua/MenuApi.cpp [hunk 3]
```diff
@@ -156,16 +156,16 @@ void LuaContext::destroy_menus() {
...
     it->recently_added = false;
     if (it->ref.is_empty()) {
       // Empty ref on a menu means that we should remove it.
       // In this case, context must also be nullptr.
       Debug::check_assertion(it->context == nullptr, "Menu with context and no ref");
-      menus.erase(it--);
+      it = menus.erase(it);
+    } else {
+      ++it;
     }
```
Comment (christopho) [inline src/lua/MenuApi.cpp]: Nice catch!

File: src/entities/Enemy.cpp [hunk 4]
```diff
@@ -819,8 +819,8 @@ void Enemy::update() {
   }
 
   if (exploding) {
-    uint32_t now = System::now();
```
Comment (christopho) [inline src/entities/Enemy.cpp]: Just remove this line, both "now" variables actually have the same value.

**Comment by Morwenn** [comment]
We never cared about `-Wshadow` until there, which probably covers C4456, C4457 and C4459. The warning tends to get really annoying when you already know you're shadowing names and it's fine. Do we really want to get rid of every name shadowing in the project? To be honest, enabling this warning alone causes hundreds of warnings in solarus.

**Comment by christopho** [comment]
Actually, this pull request contains two very unrelated fixes. I am fixing the iterator bug out of the pull request, and in the master branch. Let's keep this pull request for the warnings only.

**Comment by CitizenKyle** [comment]
OK cool. I'll just disable those warnings instead.

**Comment by Morwenn** [comment]
I think disabling these warnings is indeed the best thing to do. I have tried to compile the project with MSVC 2015 RTM too and these warnings either come from included headers we can't change or are almost nothing more than noise (seriously, C4514 is what you would expect from any library; making it a warning is rather stupid).

**Pull request closed without merging**